### PR TITLE
Insert a reveal in the default decreases clause for functions with erased arguments

### PR DESCRIPTION
**Pull request by nikswamy**
Consider:

```
let rec append #a (#m #:erased nat) (v0:vec a m) (v1:vec a n)
  : vec a (m + n)
  = match v0 with 
      | .... (* usual implementation of append here *)
```

This fails to check because the default decreases clause is the lex order on `m, n, v0, v1` and terms of type `erased _` are not related by the default precedes relation.

So, you have to write

```
 ... Tot (vec a (m + n)) (decreases (reveal m)
```

This PR enhances the default decreases clause selection procedure by inserting a reveal on any arguments of erased type in the lex order that is otherwise unchanged.

See doc/book/code/VecErased.fst for a test case.

### CODE REVIEW

File: src/typechecker/FStar.TypeChecker.TcTerm.fst [hunk 1]
```diff
@@ -462,13 +462,31 @@ let guard_letrecs env actuals expected_c : list (lbname*typ*univ_names) =
...
+                  let env = Env.push_binders env [binder] in
+                  match t.n with
+                  | Tm_type _
+                  | Tm_arrow _ -> 
+                    (out, env)
+                  | _ ->
+                    let arg = S.bv_to_name b in
+                    let arg =
+                      match is_erased_head t with
+                      | Some (u, ty) -> U.apply_reveal u ty arg
```
Comment (nikswamy) [inline src/typechecker/FStar.TypeChecker.TcTerm.fst:481]: The check is very syntactic: If the head constructor of the argument's type is `erased`, then insert a reveal in the decreases clause

**Comment by nikswamy** [comment]
Oh, this PR also adds the long-discussed

```
(* pull is an axiom.

   It can be given a model in some implementations of ``erased`.

   It says that that GTot effect is idempotent and non-dependent,
   meaning that if evaluating `f` on an argument `v:a`, exhibits an
   effect `GTot` and returns a result; then the effect does not depend
   on `v` and it can be subsumed to exhibiting the effect first and
   then computing `f v` purely.

   In other words, it "pulls" the effect out of `f`.

   pull is useful to mimic a kind of Tot/GTotc effect polymorphism.

   E.g.,  if you have `f: a -> GTot b` and a `l:list a`
          you can do List.map (pull f) l : GTot (list b)
 *)
val pull (#a #b:Type) (f: a -> GTot b) : GTot (g:(a -> b) { forall x. f x == g x })
```

**Comment by nikswamy** [comment]
Rephrased pull in a separate module FStar.Ghost.Pull using a lemma with an SMT pattern. This isolates the pull axiom and advoids clients needlessly, or mistakenly, relying on it. Also, the pattern better controls how the SMT solver is allowed to reason about the equivalence of pull.
